### ISSUE TITLE
usb-gadget bugs

### ISSUE THREAD

**Issue by nazya** [rvaiya/keyd#726]
Hi! My daily config wouldn't work in the usb-gadget mode. Here are some examples.

sbc -- machine running keyd
host -- machine using sbc as a keyboard

1. [main].leftshift = overload(shift, comma) does not produce the macro.

2. [main].a = timeout(oneshotm(typing, a), 300, 1) :  when 'a' key normally pressed and released, the host machine does not receive release event. If  several keys pressed in a stroke (when 'a' key is still being pressed) release event could be received sometimes.

3.  [rightalt+space].a = 1 does not work when normally pressed and released. The behavior is similar to the case 2, e.g. when [rightalt+space].a = 1 is pressed in sequence with [rightalt+space].s = 2, up and down events could sometimes be monitored by keyd -m on the host machine.

It seems that the problem caused by macro, and my impression is that some delay between up and down events could solve the problem.

@rvaiya @gkbd could you please comment on that? Where could be the problem? I'll do my best to fix it.

Here the example config:

[ids]
0001:0001

[global]
oneshot_timeout = 300
overload_tap_timeout = 300

[main]
leftshift = overload(shift, comma)
rightshift = overload(shift, .)
space = overload(space, space)
rightalt = overload(rightalt, enter)
rightcontrol = layer(rightalt)
a = timeout(oneshotm(typing, macro(a)), 300, 1)

[typing]
a = oneshotm(typing, macro(a))

[rightalt:C]
a = C-a

[space]
k = left

[rightalt+space]
a = 1
s = 2

**Comment by rvaiya** [comment]
Unfortunately I still haven't acquired an SBC, so @gkbd remains the sole authority on this code. 

Assuming there isn't a bug in the vkbd implementation, macros being implicated would indeed be consistent with timing issues. You may want to try experimenting with `macro_sequence_timeout` if you haven't already. Running keyd with KEYD_DEBUG=1 should also provide some clues.

**Comment by nazya** [comment]
Thanks @rvaiya ! I've just fixed the issue with introducing a 10ms delay in send_hid_report. It seems that HID mode possesses a narrow bandwidth. I will check if it is possible to dynamically set the necessary delay. Should I close the issue?

Now I'm running keyd on orange pi zero 2w, but my plan is to compile keyd for something smaller and more power efficient, like Ox64 or luckfox pico. I also dream about porting keyd on raspberry pi pico to have a more functional alternative to qmk, since porting seems easier than implementing all the missing functionality for qmk. Would appreciate your thoughts on that.

**Comment by gkbd** [comment]
Slow USB gadget might be spectific to Allwinner devices, according to
https://linux-sunxi.org/Linux_mainlining_effort#Drivers_that_can_still_be_improved.2Fadded
- sunxi-musb driver lacks DMA support (with current driver, USB gadgets are limited to PIO, limiting speed to 10MiB/s and causing large CPU)

I have a few sunxi divices but did not encounter issue myself, maybe because I was mostly testing simple keymaps.

Ox64 has only one configurable USB port, second one is for power only.
Lichee RV dock on the other hand has both host and otg,
but I did not properly test it yet, I should still have device somewhere and might try it one day. It does take long time to boot though.

Porting to RP2040 might be challenging, there is hid-remapper that runs on those MCU boards, but it's not as versatile as keyd:
https://github.com/jfedor2/hid-remapper

The advantage of microcontrollers is much faster boot and much lower power requirements. I tried couple of RISC-V boards in Kite project,
it's quick proof of the concept limited to simple maps
	https://lykt.xyz/kite/
porting keyd to those devices would require more work.
